File: NET/Libraries/Treehopper.Libraries/Displays/CharacterDisplay.cs
using System.Collections.ObjectModel;
using System.Text;
using System.Threading.Tasks;

namespace Treehopper.Libraries.Displays
{
    /// <summary>
    ///     Represents displays that can print strings of alphanumeric characters on one or more lines
    /// </summary>
    public abstract class CharacterDisplay : ICharacterDisplay
    {
        private int cursorLeft;

        private int cursorTop;

        /// <summary>
        ///     Construct a character display
        /// </summary>
        /// <param name="Columns">Number of columns</param>
        /// <param name="Rows">Number of rows</param>
        public CharacterDisplay(int Columns, int Rows)
        {
            this.Columns = Columns;
            this.Rows = Rows;
            for (var i = 0; i < Rows; i++)
                Lines.Add("");
        }

        /// <summary>
        ///     History of previously-written lines from the display
        /// </summary>
        public Collection<string> History { get; } = new Collection<string>();

        /// <summary>
        ///     Current lines of the display
        /// </summary>
        public Collection<string> Lines { get; } = new Collection<string>();

        /// <summary>
        ///     Number of columns of the display
        /// </summary>
        public int Columns { get; }

        /// <summary>
        ///     Number of rows of the display
        /// </summary>
        public int Rows { get; }

        /// <summary>
        ///     Cursor left position
        /// </summary>
        public int CursorLeft
        {
            get { return cursorLeft; }
            set
            {
                if (cursorLeft == value) return;
                cursorLeft = value;
                updateCursorPosition().Wait();
            }
        }

        /// <summary>
        ///     Cursor top position
        /// </summary>
        public int CursorTop
        {
            get { return cursorTop; }
            set
            {
                if (cursorTop == value) return;
                cursorTop = value;
                updateCursorPosition().Wait();
            }
        }

        /// <summary>
        ///     Set left/right cursor position
        /// </summary>
        /// <param name="left">The left position</param>
        /// <param name="top">The top position</param>
        /// <returns>An awaitable task that completes once the cursor is updated</returns>
        public Task SetCursorPosition(int left, int top)
        {
            cursorLeft = left;
            cursorTop = top;
            return updateCursorPosition();
        }

        /// <summary>
        ///     Write a line of text, advancing the cursor to the next line
        /// </summary>
        /// <param name="value">the text to write</param>
        /// <returns>An awaitable task that completes when finished</returns>
        public Task WriteLine(dynamic value)
        {
            return Write(value + "\n");
        }

        /// <summary>
        ///     Write text
        /// </summary>
        /// <param name="value">the text to write</param>
        /// <returns>An awaitable task that completes when finished</returns>
        public async Task Write(dynamic value)
        {
            if (cursorTop > Rows - 1)
            {
                // we need to shift all the text up, so just clear the display and resend
                // from history
                await Clear().ConfigureAwait(false);
                var startingRow = History.Count - Rows + 1;
                for (var i = 0; i < Rows - 1; i++)
                {
                    await write(History[startingRow + i]).ConfigureAwait(false);
                    await SetCursorPosition(0, i + 1).ConfigureAwait(false);
                }
            }
            string str = value.ToString();
            var s = new StringBuilder();
            foreach (var c in str)
                if (c == '\n')
                {
                    if (s.Length > 0)
                    {
                        History.Add(s.ToString());
                        await write(s).ConfigureAwait(false);
                    }
                    s.Clear();
                    await SetCursorPosition(0, cursorTop + 1).ConfigureAwait(false);
                }
                else
                {
                    s.Append(c);
                }
            if (s.Length > 0)
            {
                History.Add(s.ToString());
                await write(s).ConfigureAwait(false);
            }
        }

        /// <summary>
        ///     Clear the display and reset the cursor to (0,0)
        /// </summary>
        /// <returns>An awaitable task that completes when finished</returns>
        public Task Clear()
        {
            cursorLeft = 0;
            cursorTop = 0;
            return clear();
        }

        // Methods that display drivers need to implement

        /// <summary>
        ///     Clear the display. The driver is expected to reset the cursor to home
        /// </summary>
        protected abstract Task clear();

        /// <summary>
        ///     internally updates the current cursor position
        /// </summary>
        /// <returns></returns>
        protected abstract Task updateCursorPosition();

        /// <summary>
        ///     write text at the current cursor position
        /// </summary>
        /// <param name="value">the data to write</param>
        /// <returns>An awaitable task that completes when finished</returns>
        protected abstract Task write(dynamic value);
    }
}

File: Python/treehopper/libraries/displays/CharacterDisplay.py
from treehopper.libraries.Flushable import Flushable
from abc import ABC, abstractmethod

class CharacterDisplay(Flushable):
    def __init__(self, columns: int, rows: int):
        self.cols = columns  # type: int
        self.rows = rows     # type: rows
        self.lines = []      # type list[string]
        self._cursor_left = 0
        self._cursor_top = 0

        for i in range(rows):
            self.lines.append("")

    def flush(self, force=False):
        pass

    @property
    def cursor_left(self):
        return self._cursor_left

    @cursor_left.setter
    def cursor_left(self, value):
        if self._cursor_left == value:
            return

        self._cursor_left = value

        # check for overflow
        if self._cursor_left >= self.cols:
            self._cursor_left = 0

        self._update_cursor()

    @property
    def cursor_top(self):
        return self._cursor_top

    @cursor_top.setter
    def cursor_top(self, value):
        if self._cursor_top == value:
            return

        self._cursor_top = value

        # check for overflow
        if self._cursor_top >= self.rows:
            self._cursor_top = 0

        self._update_cursor()

    def set_cursor(self, left: int, top: int):
        self._cursor_left = left
        self._cursor_top = top
        self._update_cursor()

    def clear(self):
        self._cursor_left = 0
        self._cursor_top = 0
        self._clear()

    def write(self, value):
        string = str(value)
        for c in string:
            self._write_char(c)
            self.cursor_left += 1
        self.flush()

    @abstractmethod
    def _clear(self):
        pass

    @abstractmethod
    def _update_cursor(self):
        pass

    @abstractmethod
    def _write_char(self, character):
        pass

    @abstractmethod
    def flush(self, force=False):
        pass
